sway-core/src/asm_generation/fuel/optimizations.rs

```diff
@@ -0,0 +1,257 @@
+                        if let Some(c) = data_section.get_data_word(data_id) {
+                            reg_contents.insert(dest.clone(), RegContents::Constant(c));
+                        } else {
+                            reg_contents.remove(dest);
+                        }
+                        record_new_def(&mut latest_version, dest);
+                    }
+                    VirtualOp::MOVI(dest, imm) => {
+                        reg_contents.insert(dest.clone(), RegContents::Constant(imm.value as u64));
+                        record_new_def(&mut latest_version, dest);
+                    }
+                    VirtualOp::LW(dest, addr_reg, imm) => match reg_contents.get(addr_reg) {
+                        Some(RegContents::BaseOffset(base_reg, offset))
+                            if get_def_version(&latest_version, &base_reg.reg) == base_reg.ver
+                                && ((offset / 8) + imm.value as u64)
+                                    < compiler_constants::TWELVE_BITS =>
+                        {
+                            let new_imm = VirtualImmediate12::new_unchecked(
+                                (offset / 8) + imm.value as u64,
+                                "Immediate offset too big for LW",
+                            );
+                            let new_lw = VirtualOp::LW(dest.clone(), base_reg.reg.clone(), new_imm);
+                            // The register defined is no more useful for us. Forget anything from its past.
+                            reg_contents.remove(dest);
+                            record_new_def(&mut latest_version, dest);
+                            // Replace the LW with a new one in-place.
+                            *op = new_lw;
+                        }
+                        _ => {
+                            reg_contents.remove(dest);
+                            record_new_def(&mut latest_version, dest);
+                        }
+                    },
+                    VirtualOp::SW(addr_reg, src, imm) => match reg_contents.get(addr_reg) {
+                        Some(RegContents::BaseOffset(base_reg, offset))
+                            if get_def_version(&latest_version, &base_reg.reg) == base_reg.ver
+                                && ((offset / 8) + imm.value as u64)
+                                    < compiler_constants::TWELVE_BITS =>
+                        {
+                            let new_imm = VirtualImmediate12::new_unchecked(
+                                (offset / 8) + imm.value as u64,
+                                "Immediate offset too big for SW",
+                            );
+                            let new_sw = VirtualOp::SW(base_reg.reg.clone(), src.clone(), new_imm);
+                            // Replace the SW with a new one in-place.
+                            *op = new_sw;
+                        }
+                        _ => (),
```

Comment:
Same here, shouldn't we remove the `dest` and record the new definition?

If there is some reason that justifies this different behavior between `SW` and `LW` I don't see it at the first look :smile: In that case perhaps adding a comment would be great.

Reply:
Answered above.